test/OpenTelemetry.AotCompatibility.Tests/AotCompatibilityTests.cs

```diff
@@ -0,0 +1,78 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System.Diagnostics;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace OpenTelemetry.AotCompatibility.Tests
+{
+    public class AotCompatibilityTests
+    {
+        private readonly ITestOutputHelper testOutputHelper;
+
+        public AotCompatibilityTests(ITestOutputHelper testOutputHelper)
+        {
+            this.testOutputHelper = testOutputHelper;
+        }
+
+        /// <summary>
+        /// This test ensures that the intended APIs of the OpenTelemetry.AotCompatibility.TestApp libraries are
+        /// trimming and NativeAOT compatible.
+        ///
+        /// This test follows the instructions in https://learn.microsoft.com/dotnet/core/deploying/trimming/prepare-libraries-for-trimming#show-all-warnings-with-sample-application
+        ///
+        /// If this test fails, it is due to adding trimming and/or AOT incompatible changes
+        /// to code that is supposed to be compatible.
+        ///
+        /// To diagnose the problem, inspect the test output which will contain the trimming and AOT errors. For example:
+        ///
+        /// error IL2091: 'T' generic argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicConstructors'.
+        /// </summary>
+        [Fact]
+        public void EnsureAotCompatibility()
+        {
+            string testAppPath = @"..\..\..\..\OpenTelemetry.AotCompatibility.TestApp";
+            string testAppProject = "OpenTelemetry.AotCompatibility.TestApp.csproj";
+
+            // ensure we run a clean publish every time
+            DirectoryInfo testObjDir = new DirectoryInfo(Path.Combine(testAppPath, "obj"));
+            if (testObjDir.Exists)
+            {
+                testObjDir.Delete(recursive: true);
+            }
+
+            var process = new Process
+            {
+                // set '-nodereuse:false /p:UseSharedCompilation=false' so the MSBuild and Roslyn server processes don't hang around, which may hang the test in CI
+                StartInfo = new ProcessStartInfo("dotnet", $"publish {testAppProject} --self-contained -nodereuse:false /p:UseSharedCompilation=false")
+                {
+                    RedirectStandardOutput = true,
+                    UseShellExecute = false,
+                    CreateNoWindow = true,
+                    WorkingDirectory = testAppPath,
+                },
+            };
+            process.OutputDataReceived += (sender, e) => this.testOutputHelper.WriteLine(e.Data);
+            process.Start();
+            process.BeginOutputReadLine();
+
+            Assert.True(process.WaitForExit(milliseconds: 180_000), "dotnet publish command timed out after 180 seconds.");
+
+            Assert.True(process.ExitCode == 0, "Publishing the AotCompatibility app failed. See test output for more details.");
```

Comment:
You may want to consider a way to baseline the existing warnings to start, and ensure new warnings aren't introduced as new features are added. I'm not sure the best way to do this, but maybe @vitek-karas @agocke or @sbomer knows of a good way?

In dotnet/sdk, we use the following method to verify the warnings (see the callers above this method):

https://github.com/dotnet/sdk/blob/7d23e9d3e4aad58a5b497d8d91a50ffdf148b238/src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToRunILLink.cs#L803-L834

Reply:
Other than baselining the output of the build I can't think of another way for NativeAOT right now.

ILLink supports suppressing warnings via XML files, which would be the best solution here, unfortunately that functionality isn't implemented in NativeAOT (it's on the list, but I don't know when it's going to happen).
NativeAOT does support suppressing warnings in code via assembly-level attributes, but those would have to go onto the assembly which produces the warnings, which is not useful to this case.